In [1]:
import numpy as np
import pandas as pd

In [5]:
df = pd.read_csv(r"C:/Users/Lenovo/Desktop/Pandas Practice/UCI Online Retail Dataset/data.csv", encoding='unicode_escape')

In [6]:
df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


In [7]:
df.shape

(541909, 8)

In [8]:
df.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


In [14]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [16]:
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

In [17]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

# It is noticed that customer id is in float, so we converted it to int32

In [29]:
df['CustomerID'] = df['CustomerID'].astype('Int32')
df['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: Int32

In [30]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850,United Kingdom


In [31]:
df.InvoiceNo.value_counts()

573585     1114
581219      749
581492      731
580729      721
558475      705
           ... 
554023        1
554022        1
554021        1
554020        1
C558901       1
Name: InvoiceNo, Length: 25900, dtype: int64

# It is noticed that some invoice numbers appears several times above |

In [32]:
df.groupby('InvoiceNo')['Description'].value_counts()

InvoiceNo  Description                        
536365     CREAM CUPID HEARTS COAT HANGER         1
           GLASS STAR FROSTED T-LIGHT HOLDER      1
           KNITTED UNION FLAG HOT WATER BOTTLE    1
           RED WOOLLY HOTTIE WHITE HEART.         1
           SET 7 BABUSHKA NESTING BOXES           1
                                                 ..
C581490    ZINC T-LIGHT HOLDER STARS SMALL        1
C581499    Manual                                 1
C581568    VICTORIAN SEWING BOX LARGE             1
C581569    36 PENCILS TUBE RED RETROSPOT          1
           HANGING HEART JAR T-LIGHT HOLDER       1
Name: Description, Length: 529586, dtype: int64

# above we see that several invoice numbers have C as prefix which means cancelled invoice

# Lets create a separate dataframe for Cancelled invoice numbers

In [38]:
cancelled_invoice_df = df[df['InvoiceNo'].astype('str').str.startswith('C')]

In [39]:
cancelled_invoice_df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315,United Kingdom


# we notice that entries where unit price is 0, have no customer id. These look like data entry errors and should be excluded

In [40]:
df[df['UnitPrice'] == 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,<NA>,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,<NA>,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,<NA>,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,<NA>,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,<NA>,United Kingdom
...,...,...,...,...,...,...,...,...
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,<NA>,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,12/8/2011 13:58,0.0,<NA>,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,12/8/2011 13:58,0.0,<NA>,United Kingdom
538554,581408,85175,NaN,20,12/8/2011 14:06,0.0,<NA>,United Kingdom


In [41]:
# We find the index labels of the rows where price is 0, then drop them
df.drop(df[df['UnitPrice'] == 0].index, inplace=True)

In [42]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850,United Kingdom


In [43]:
df.shape

(539394, 8)

# Adding a TotalRevenue column to dataframe--

In [44]:
df['TotalRevenue'] = df['UnitPrice'] * df['Quantity']

In [45]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalRevenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850,United Kingdom,20.34


# We also noticed that 135080 customers have no customer id

# Let's Calculate what percentage of total revenue is "anonymous" versus "registered."

In [53]:
is_anonymous = df['CustomerID'].isnull()

In [56]:
anonymous_revenue = df[is_anonymous]['TotalRevenue'].sum()
anonymous_revenue

1447682.1200000003

In [57]:
registered_revenue = df[~is_anonymous]['TotalRevenue'].sum()
registered_revenue

8300065.814

In [60]:
total_revenue = df['TotalRevenue'].sum()
total_revenue

9747747.933999998

# % of anonymous revenue

In [63]:
print(round((anonymous_revenue/total_revenue) * 100,2),'%')

14.85 %


# % of registered revenue

In [64]:
print(round((registered_revenue/total_revenue) * 100,2),'%')

85.15 %


# Cohort Analysis: For each customer, find their first purchase date (Month/Year). Then, calculate how many customers from each "cohort" returned to buy something in subsequent months.

In [65]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalRevenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850,United Kingdom,20.34


In [68]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [70]:
df['InvoiceMonth'] = df['InvoiceDate'].dt.to_period('M')

In [71]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalRevenue,InvoiceMonth
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30,2010-12
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00,2010-12
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12


In [72]:
# Transform finds the min date for each group and maps it back to the original index
df['CohortMonth'] = df.groupby('CustomerID')['InvoiceMonth'].transform('min')

In [74]:
# Calculate difference in months
def get_date_int(df, column):
    year = df[column].dt.year
    month = df[column].dt.month
    return year, month

invoice_year, invoice_month = get_date_int(df, 'InvoiceMonth')
cohort_year, cohort_month = get_date_int(df, 'CohortMonth')

years_diff = invoice_year - cohort_year
months_diff = invoice_month - cohort_month

# Calculate total months difference
df['CohortIndex'] = years_diff * 12 + months_diff

In [80]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalRevenue,InvoiceMonth,CohortMonth,CohortIndex
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30,2010-12,2010-12,0
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00,2010-12,2010-12,0
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0


In [75]:
# Group by Cohort and the Index, then count unique customers
cohort_data = df.groupby(['CohortMonth', 'CohortIndex'])['CustomerID'].nunique().reset_index()

# Pivot the table so Cohorts are rows and Months passed are columns
cohort_counts = cohort_data.pivot(index='CohortMonth', columns='CohortIndex', values='CustomerID')

# Convert to percentages (Retention Rate)
cohort_sizes = cohort_counts.iloc[:, 0]
retention = cohort_counts.divide(cohort_sizes, axis=0) * 100

In [81]:
cohort_data

,CohortMonth,CohortIndex,CustomerID
0,2010-12,0,948
1,2010-12,1,362
2,2010-12,2,317
3,2010-12,3,367
4,2010-12,4,341
...,...,...,...
86,2011-10,1,93
87,2011-10,2,46
88,2011-11,0,320
89,2011-11,1,43


In [79]:
cohort_counts

CohortIndex,0,1,2,3,4,5,6,7,8,9,10,11,12
CohortMonth,,,,,,,,,,,,,
2010-12,948.0,362.0,317.0,367.0,341.0,376.0,360.0,336.0,336.0,374.0,354.0,474.0,260.0
2011-01,421.0,101.0,119.0,102.0,138.0,126.0,110.0,108.0,131.0,146.0,155.0,63.0,NaN
2011-02,380.0,94.0,73.0,106.0,102.0,94.0,97.0,107.0,98.0,119.0,35.0,NaN,NaN
2011-03,440.0,84.0,112.0,96.0,102.0,78.0,116.0,105.0,127.0,39.0,NaN,NaN,NaN
2011-04,299.0,68.0,66.0,63.0,62.0,71.0,69.0,78.0,25.0,NaN,NaN,NaN,NaN
2011-05,279.0,66.0,48.0,48.0,60.0,68.0,74.0,29.0,NaN,NaN,NaN,NaN,NaN
2011-06,235.0,49.0,44.0,64.0,58.0,79.0,24.0,NaN,NaN,NaN,NaN,NaN,NaN
2011-07,191.0,40.0,39.0,44.0,52.0,22.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-08,167.0,42.0,42.0,42.0,23.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [82]:
cohort_sizes

CohortMonth
2010-12    948.0
2011-01    421.0
2011-02    380.0
2011-03    440.0
2011-04    299.0
2011-05    279.0
2011-06    235.0
2011-07    191.0
2011-08    167.0
2011-09    298.0
2011-10    352.0
2011-11    320.0
2011-12     41.0
Freq: M, Name: 0, dtype: float64

In [83]:
retention

CohortIndex,0,1,2,3,4,5,6,7,8,9,10,11,12
CohortMonth,,,,,,,,,,,,,
2010-12,100.0,38.185654,33.438819,38.713080,35.970464,39.662447,37.974684,35.443038,35.443038,39.451477,37.341772,50.000000,27.42616
2011-01,100.0,23.990499,28.266033,24.228029,32.779097,29.928741,26.128266,25.653207,31.116390,34.679335,36.817102,14.964371,NaN
2011-02,100.0,24.736842,19.210526,27.894737,26.842105,24.736842,25.526316,28.157895,25.789474,31.315789,9.210526,NaN,NaN
2011-03,100.0,19.090909,25.454545,21.818182,23.181818,17.727273,26.363636,23.863636,28.863636,8.863636,NaN,NaN,NaN
2011-04,100.0,22.742475,22.073579,21.070234,20.735786,23.745819,23.076923,26.086957,8.361204,NaN,NaN,NaN,NaN
2011-05,100.0,23.655914,17.204301,17.204301,21.505376,24.372760,26.523297,10.394265,NaN,NaN,NaN,NaN,NaN
2011-06,100.0,20.851064,18.723404,27.234043,24.680851,33.617021,10.212766,NaN,NaN,NaN,NaN,NaN,NaN
2011-07,100.0,20.942408,20.418848,23.036649,27.225131,11.518325,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-08,100.0,25.149701,25.149701,25.149701,13.772455,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [84]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalRevenue,InvoiceMonth,CohortMonth,CohortIndex
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30,2010-12,2010-12,0
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00,2010-12,2010-12,0
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0


In [90]:
snapshot_date = df['InvoiceDate'].max() + pd.Timedelta(days=1)

In [95]:
rfm = df.groupby('CustomerID').agg({
    'InvoiceDate': lambda x: (snapshot_date - x.max()).days,
    'InvoiceNo': 'nunique',
    'TotalRevenue': 'sum'
}
)

In [96]:
rfm

,InvoiceDate,InvoiceNo,TotalRevenue
CustomerID,,,
12346,326,2,0.00
12347,2,7,4310.00
12348,75,4,1797.24
12349,19,1,1757.55
12350,310,1,334.40
...,...,...,...
18280,278,1,180.60
18281,181,1,80.82
18282,8,3,176.60


In [101]:
# Rename columns for clarity
rfm.rename(columns={
    'InvoiceDate': 'Recency',
    'InvoiceNo': 'Frequency',
    'TotalRevenue': 'Monetary'
}, inplace=True)

In [102]:
rfm

,Recency,Frequency,Monetary,R,F
CustomerID,,,,,
12346,326,2,0.00,1,2
12347,2,7,4310.00,4,4
12348,75,4,1797.24,2,3
12349,19,1,1757.55,3,1
12350,310,1,334.40,1,1
...,...,...,...,...,...
18280,278,1,180.60,1,2
18281,181,1,80.82,1,2
18282,8,3,176.60,4,3


In [103]:
# Quick summary of the RFM distribution
print(rfm.describe())

           Recency    Frequency       Monetary
count  4371.000000  4371.000000    4371.000000
mean     92.064974     5.075726    1898.894032
std     100.770046     9.332529    8220.235339
min       1.000000     1.000000   -4287.630000
25%      17.000000     1.000000     293.785000
50%      50.000000     3.000000     648.410000
75%     143.000000     5.000000    1611.860000
max     374.000000   248.000000  279489.020000


In [105]:
# Example: Dividing Recency into 4 quantiles (1 is best/most recent)
r_labels = range(4, 0, -1)
rfm['R'] = pd.qcut(rfm['Recency'], q=4, labels=r_labels)

# Example: Dividing Frequency and Monetary (4 is best)
f_labels = range(1, 5)
m_labels = range(1, 5)
rfm['F'] = pd.qcut(rfm['Frequency'].rank(method='first'), q=4, labels=f_labels)
rfm['M'] = pd.qcut(rfm['Monetary'], q=4, labels=m_labels)

In [106]:
rfm

,Recency,Frequency,Monetary,R,F,M
CustomerID,,,,,,
12346,326,2,0.00,1,2,1
12347,2,7,4310.00,4,4,4
12348,75,4,1797.24,2,3,4
12349,19,1,1757.55,3,1,4
12350,310,1,334.40,1,1,2
...,...,...,...,...,...,...
18280,278,1,180.60,1,2,1
18281,181,1,80.82,1,2,1
18282,8,3,176.60,4,3,1


# Top Selling Products by Country: Find the top 3 most sold products (by quantity) for every country

In [107]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalRevenue,InvoiceMonth,CohortMonth,CohortIndex
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30,2010-12,2010-12,0
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00,2010-12,2010-12,0
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0


In [114]:
country_product_totals = df.groupby(['Country', 'Description'])['Quantity'].sum().reset_index()
country_product_totals

,Country,Description,Quantity
0,Australia,DOLLY GIRL BEAKER,200
1,Australia,I LOVE LONDON MINI BACKPACK,4
2,Australia,10 COLOUR SPACEBOY PEN,48
3,Australia,12 PENCIL SMALL TUBE WOODLAND,384
4,Australia,12 PENCILS TALL TUBE POSY,252
...,...,...,...
19987,Unspecified,WRAP POPPIES DESIGN,25
19988,Unspecified,WRAP SUKI AND FRIENDS,25
19989,Unspecified,WRAP VINTAGE PETALS DESIGN,25
19990,Unspecified,WRAP WEDDING DAY,25


In [120]:
top_3_per_country = country_product_totals.groupby('Country').apply(
    lambda x: x.nlargest(3, 'Quantity')
).reset_index(drop=True)

top_3_per_country

,Country,Description,Quantity
0,Australia,MINI PAINT SET VINTAGE,2916
1,Australia,RABBIT NIGHT LIGHT,1884
2,Australia,RED HARMONICA IN BOX,1704
3,Austria,SET 12 KIDS COLOUR CHALK STICKS,288
4,Austria,MINI JIGSAW PURDEY,240
...,...,...,...
109,United Kingdom,JUMBO BAG RED RETROSPOT,43163
110,United Kingdom,POPCORN HOLDER,34365
111,Unspecified,WORLD WAR 2 GLIDERS ASSTD DESIGNS,96
112,Unspecified,72 SWEETHEART FAIRY CAKE CASES,48


In [124]:
print(top_3_per_country[top_3_per_country['Country'].isin(['United Kingdom', 'Germany'])].sort_values(by=['Country', 'Quantity'], ascending=[True, False]))

            Country                          Description  Quantity
42          Germany  ROUND SNACK BOXES SET OF4 WOODLAND       1218
43          Germany            ASSORTED COLOURS SILK FAN      1164
44          Germany                              POSTAGE      1104
108  United Kingdom    WORLD WAR 2 GLIDERS ASSTD DESIGNS     48326
109  United Kingdom              JUMBO BAG RED RETROSPOT     43163
110  United Kingdom                       POPCORN HOLDER     34365


In [122]:
top_3_per_country.sort_values(by=['Country', 'Quantity'], ascending=[True, False])

,Country,Description,Quantity
0,Australia,MINI PAINT SET VINTAGE,2916
1,Australia,RABBIT NIGHT LIGHT,1884
2,Australia,RED HARMONICA IN BOX,1704
3,Austria,SET 12 KIDS COLOUR CHALK STICKS,288
4,Austria,MINI JIGSAW PURDEY,240
...,...,...,...
109,United Kingdom,JUMBO BAG RED RETROSPOT,43163
110,United Kingdom,POPCORN HOLDER,34365
111,Unspecified,WORLD WAR 2 GLIDERS ASSTD DESIGNS,96
112,Unspecified,72 SWEETHEART FAIRY CAKE CASES,48


In [125]:
# Filter first, then sort globally by Quantity
filtered_df = top_3_per_country[top_3_per_country['Country'].isin(['United Kingdom', 'Germany'])]
print(filtered_df.sort_values(by='Quantity', ascending=False))

            Country                          Description  Quantity
108  United Kingdom    WORLD WAR 2 GLIDERS ASSTD DESIGNS     48326
109  United Kingdom              JUMBO BAG RED RETROSPOT     43163
110  United Kingdom                       POPCORN HOLDER     34365
42          Germany  ROUND SNACK BOXES SET OF4 WOODLAND       1218
43          Germany            ASSORTED COLOURS SILK FAN      1164
44          Germany                              POSTAGE      1104


# Resampling for Growth: Resample the data to a "Weekly" frequency and calculate the Week-over-Week (WoW) growth in revenue.

In [129]:
df_time = df.set_index('InvoiceDate')
df_time.head()

,InvoiceNo,StockCode,Description,Quantity,UnitPrice,CustomerID,Country,TotalRevenue,InvoiceMonth,CohortMonth,CohortIndex
InvoiceDate,,,,,,,,,,,
2010-12-01 08:26:00,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2.55,17850,United Kingdom,15.30,2010-12,2010-12,0
2010-12-01 08:26:00,536365,71053,WHITE METAL LANTERN,6,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
2010-12-01 08:26:00,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2.75,17850,United Kingdom,22.00,2010-12,2010-12,0
2010-12-01 08:26:00,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0
2010-12-01 08:26:00,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,3.39,17850,United Kingdom,20.34,2010-12,2010-12,0


In [132]:
weekly_revenue = df_time['TotalRevenue'].resample('W').sum()
weekly_revenue

InvoiceDate
2010-12-05    181847.250
2010-12-12    270287.030
2010-12-19    207052.170
2010-12-26     89770.570
2011-01-02         0.000
2011-01-09     93720.930
2011-01-16    190994.960
2011-01-23    133782.910
2011-01-30    119136.810
2011-02-06    123642.180
2011-02-13    102296.930
2011-02-20    139664.580
2011-02-27    133069.930
2011-03-06    130435.560
2011-03-13    129180.960
2011-03-20    145262.610
2011-03-27    147909.800
2011-04-03    183501.710
2011-04-10    121598.220
2011-04-17    147459.771
2011-04-24    115461.250
2011-05-01     84382.660
2011-05-08    130354.310
2011-05-15    205445.230
2011-05-22    198031.530
2011-05-29    160569.820
2011-06-05    116932.520
2011-06-12    177916.870
2011-06-19    184016.500
2011-06-26    117086.710
2011-07-03    136287.440
2011-07-10    174659.170
2011-07-17    124163.170
2011-07-24    182513.960
2011-07-31    180814.851
2011-08-07    153176.200
2011-08-14    161417.020
2011-08-21    170724.890
2011-08-28    149603.500
2011-09-04   

In [135]:
wow_growth = weekly_revenue.pct_change() * 100
wow_growth

InvoiceDate
2010-12-05           NaN
2010-12-12     48.634104
2010-12-19    -23.395447
2010-12-26    -56.643502
2011-01-02   -100.000000
2011-01-09           inf
2011-01-16    103.791149
2011-01-23    -29.954743
2011-01-30    -10.947661
2011-02-06      3.781678
2011-02-13    -17.263728
2011-02-20     36.528613
2011-02-27     -4.721777
2011-03-06     -1.979688
2011-03-13     -0.961854
2011-03-20     12.448932
2011-03-27      1.822348
2011-04-03     24.063253
2011-04-10    -33.734558
2011-04-17     21.268034
2011-04-24    -21.699831
2011-05-01    -26.916901
2011-05-08     54.479973
2011-05-15     57.605245
2011-05-22     -3.608602
2011-05-29    -18.917043
2011-06-05    -27.176527
2011-06-12     52.153456
2011-06-19      3.428360
2011-06-26    -36.371624
2011-07-03     16.398727
2011-07-10     28.155001
2011-07-17    -28.911165
2011-07-24     46.995248
2011-07-31     -0.930947
2011-08-07    -15.285609
2011-08-14      5.379961
2011-08-21      5.766350
2011-08-28    -12.371594
2011-09-04   

In [136]:
weekly_summary = pd.DataFrame({
    'WeeklyRevenue': weekly_revenue,
    'WoW_Growth_%': wow_growth
})

print(weekly_summary.head(10))

             WeeklyRevenue  WoW_Growth_%
InvoiceDate                             
2010-12-05       181847.25           NaN
2010-12-12       270287.03     48.634104
2010-12-19       207052.17    -23.395447
2010-12-26        89770.57    -56.643502
2011-01-02            0.00   -100.000000
2011-01-09        93720.93           inf
2011-01-16       190994.96    103.791149
2011-01-23       133782.91    -29.954743
2011-01-30       119136.81    -10.947661
2011-02-06       123642.18      3.781678


# identifying "Peak Hour" analysis to see when the most transactions happen during the day

In [137]:
# 1. Extract hour from the InvoiceDate
df['Hour'] = df['InvoiceDate'].dt.hour

# 2. Count unique transactions (InvoiceNo) per hour
# We use nunique because one order has multiple rows (line items)
peak_hours = df.groupby('Hour')['InvoiceNo'].nunique()

# 3. Find the peak hour
best_hour = peak_hours.idxmax()
print(f"The peak hour for transactions is: {best_hour}:00")

The peak hour for transactions is: 12:00
